services/sharedchannel/sync_recv.go

```diff
@@ -28,223 +28,208 @@ func (scs *Service) onReceiveSyncMessage(msg model.RemoteClusterMsg, rc *model.R
 		return fmt.Errorf("invalid sync message: %w", err)
 	}
 
-	scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceDebug, "Sync message received",
+	scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceDebug, "Batch of sync messages received",
 		mlog.String("remote", rc.DisplayName),
-		mlog.Int("count", len(syncMessages)),
+		mlog.Int("sync_msg_count", len(syncMessages)),
 	)
 
-	postErrors, lastUpdate, err := scs.processSyncMessagesViaStore(syncMessages, rc, response)
-	if err == nil {
-		if len(postErrors) > 0 {
-			response[ResponsePostErrors] = postErrors
-		}
-		response[ResponseLastUpdateAt] = lastUpdate
-	}
-	return err
+	return scs.processSyncMessages(syncMessages, rc, response)
 }
 
-//
-//   Sync message should be array of users; array of posts; array of reactions
-//
-
-//
-// Strategy 1:  use store to upsert posts and reactions.
-//
-
-func (scs *Service) processSyncMessagesViaStore(syncMessages []syncMsg, rc *model.RemoteCluster, response remotecluster.Response) (postErrors []string, lastUpdate int64, err error) {
-	postErrors = make([]string, 0)
+func (scs *Service) processSyncMessages(syncMessages []syncMsg, rc *model.RemoteCluster, response remotecluster.Response) error {
 	var channel *model.Channel
+	postErrors := make([]string, 0)
+	usersSyncd := make([]string, 0)
+	var lastSyncAt int64
+	var err error
+
 	chanToTeamMap := make(map[string]*model.Team)
 
 	for _, sm := range syncMessages {
-		scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceDebug, "Sync post received",
-			mlog.String("post_id", sm.Post.Id),
-			mlog.String("channel_id", sm.Post.ChannelId),
-			mlog.Int("reaction_count", len(sm.Reactions)))
-
-		if _, ok := chanToTeamMap[sm.Post.ChannelId]; !ok {
-			team, err2 := scs.server.GetStore().Channel().GetTeamForChannel(sm.Post.ChannelId)
-			if err2 != nil {
-				scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceError, "Error getting Team for Channel",
-					mlog.String("ChannelId", sm.Post.ChannelId),
-					mlog.String("PostId", sm.Post.Id),
-					mlog.Err(err2),
-				)
-				postErrors = append(postErrors, sm.Post.Id)
-				continue
-			}
-			chanToTeamMap[sm.Post.ChannelId] = team
-		}
 
-		sm.Post.Message = scs.processPermalinkFromRemote(sm.Post, chanToTeamMap[sm.Post.ChannelId])
+		scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceDebug, "Sync msg received",
+			mlog.String("post_id", sm.PostId),
+			mlog.String("channel_id", sm.ChannelId),
+			mlog.Int("reaction_count", len(sm.Reactions)),
+			mlog.Int("user_count", len(sm.Users)),
+			mlog.Bool("has_post", sm.Post != nil),
+		)
 
 		if channel == nil {
-			if channel, err = scs.server.GetStore().Channel().Get(sm.Post.ChannelId, true); err != nil {
+			if channel, err = scs.server.GetStore().Channel().Get(sm.ChannelId, true); err != nil {
```

Comment:
Can the channel change between loop iterations?

Reply:
I don't see how.  The snapshot of posts we're working with can only come from the same channel.  Even if a post was moved to another channel (wrangler) it will just appear deleted in a subsequent sync.

Reply:
But we are using `sm.ChannelId`. Does that mean that all sync messages are for the same channel?

Reply:
Yes.  On the sending side, an entire sync message is populated from a single GetPostsSince, which is channel specific.  It is a contract of the message that it can only contain post/reaction/user from a single channel.

Reply:
One last thing, is it possible that `sm.ChannelId` differs from `sm.Post.ChannelId`?
I am trying to understand why we use the former to get the channel and the latter to get the team.

Reply:
Ah, I see where the confusion is.  When Agniva added the perma-link processing he didn't realize there could be only one channel per batch of sync messages and he created a map of channels to teams.  My fault - I completely missed that when reviewing.  I will simply that part (remove the map) and also sanity check that `sm.Post.ChannelId == sm.ChannelId`.  I'll then make all reference to channel id use `sm.ChannelId`.

Reply:
Fixed via https://github.com/mattermost/mattermost-server/pull/16783/commits/9bace6eb1613a4e5f34f48b99b964ac0fac95bab